expertsystem/io/_pdg.py

```diff
@@ -0,0 +1,129 @@
+"""Create a `.ParticleCollection` instance from PDG info."""
+
+from typing import Optional, Tuple
+
+from particle import Particle as PdgDatabase
+from particle.particle import enums
+
+from expertsystem.data import (
+    Parity,
+    Particle,
+    ParticleCollection,
+    QuantumState,
+    Spin,
+)
+
+
+# cspell:ignore pdgid
+def __get_lepton_qn(pdg_particle: PdgDatabase) -> tuple:
+    electron_lepton_number = 0
+    muon_lepton_number = 0
+    tau_lepton_number = 0
+    if pdg_particle.pdgid.is_lepton:
+        if "e" in pdg_particle.name:
+            electron_lepton_number = pdg_particle.pdgid / abs(
+                pdg_particle.pdgid
+            )
+        if "mu" in pdg_particle.name:
+            muon_lepton_number = pdg_particle.pdgid / abs(pdg_particle.pdgid)
+        if "tau" in pdg_particle.name:
+            tau_lepton_number = pdg_particle.pdgid / abs(pdg_particle.pdgid)
+
+    return electron_lepton_number, muon_lepton_number, tau_lepton_number
+
+
+def __get_hadron_qn(
+    pdg_particle: PdgDatabase,
+) -> Tuple[int, float, int, int, int, int]:
+    baryonnumber = 0
+    spin_projection = 0.0
+    strangeness = 0
+    charmness = 0
+    bottomness = 0
+    topness = 0
+    baryonnumber = (
+        pdg_particle.pdgid
+        / abs(pdg_particle.pdgid)
+        * pdg_particle.pdgid.is_baryon
+    )
+    if pdg_particle.pdgid.is_hadron:
+        for quark in pdg_particle.quarks:
```

Comment:
Better to create a separate function, something like:
```python
def __compute_quark_numbers(quark_string: str) -> Tuple[int, int, int, int]:
    ...
```
It's still not clean, but the smaller the return tuple, the smaller the risk for bugs.

Reply:
I´m not sure what you mean, but the only thing in this function that is not calculated from the quarks is the baryon number.

Reply:
This function would only calculate `strangeness`, `charmness`, `bottomness`, and `topness` (return as a 4-tuple). So you extract this function from the `__get_hadron_qn` and let `__get_hadron_qn` only return `baryonnumber` and `spin_projection` (probably need to rename it).

The aim is just to keep functions small and reduce the size of the returned tuples.